rtio: Real-Time Stream and Sensor API

teburd opened the pull request:
A DMA friendly Stream API for zephyr.
    
Provides a way to define a driver which provides an output FIFO of
blocks. Blocks are allocated from a specified block allocator. These blocks are
easily used for DMA transfers.
    
Some convenience functions for storing and retrieving primitive types
in the blocks is provided.
    
Blocks are meant to be placed into the output fifo based on a policy
describing time or number of entries since the block began being filled. If
the block is full it is meant to be immediately pushed into the FIFO.
    
The use case in mind is high rate sensors with hardware FIFOs and SPI
interfaces where DMA transfers from these sensors would save a great
deal of CPU time avoiding the high rate of interrupts and small SPI
transfers that would otherwise occur. Other uses are certainly possible,
I imagine audio and video have similar requirements.

For sensors an added interface to iterate over sample sets from a block is provided allowing for lazily converting sample set values into numerical values of varying types. To start only float is available but certainly the current sensor_value (int/int pair), cmsis-dsp friendly q value types, or other numerical formats could be provided.

teburd committed:
rtio: Real-Time Stream API

A DMA friendly Stream API for zephyr.

Provides a way to define a driver which provides an outpuf fifo of
blocks. Blocks are allocated from a specified block allocator. These
blocks are easily used for DMA transfers.

Some convience functions for storing and retrieving primitive types
in the blocks is provided.

Blocks are meant to be placed into the output fifo based on a policy
describing time or number of entries since the block began being filled.
If the block is full it is meant to be immediately pushed into the fifo.

The use case in mind is high rate sensors with hardware FIFOs and SPI
interfaces where DMA transfers from these sensors would save a great
deal of CPU time avoiding the high rate of interrupts and small SPI
transfers that would otherwise occur. Other uses are certainly possible,
I imagine audio and video have similar requirements.

Signed-off-by: Tom Burdick <tom.burdick@electromatic.us>
Co-authored-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

dbkinder reviewed (changes requested):
Doesn't this new rtio API need documentation (as ZIO previously did)?

bfrog commented:
@dbkinder it does, though I'm still working out how things work, so far everyone involved has been ok with this situation while its in the topic branch not yet in master, when the PR for master happens it'll have docs

bfrog commented:
@dbkinder I added some initial documentation this time as I think its important and you are right, I will continue to add to it as this progresses

dbkinder commented:
Alas, you've tripped on a known issue in our API documentation processing around nested unnamed unions in a struct (don't ask).  You can document this as a "known issue" (and remove the reported documentation error) by adding a file to your PR ``.known-issues/doc/rtio.conf`` in the zephyr base directory, with this content (you'll find some other of these .conf files there for other "known issues"):
```
#
# RTIO
#
#
# include/rtio.h warnings
#
^(?P<filename>([\-:\\/\w\.])+[/\\]doc[/\\]reference[/\\]peripherals[/\\]rtio.rst):(?P<lineno>[0-9]+): WARNING: Invalid definition: Expected identifier in nested name. \[error at [0-9]+]
^[ \t]*
^[ \t]*\^
^(?P=filename):(?P=lineno): WARNING: Invalid definition: Expected identifier in nested name. \[error at [0-9]+]
^[ \t]*
^[ \t]*\^
^(?P=filename):(?P=lineno): WARNING: Invalid definition: Expected end of definition. \[error at [0-9]+]
^.*rtio_trigger_config.options
^[- \t]*\^
#
```

bfrog commented:
@dbkinder I believe everything is sorted now

dbkinder commented on doc/reference/peripherals/rtio.rst:
```diff
@@ -0,0 +1,49 @@
+Real-Time IO
+############
+
+The real-time io or rtio subsystem exposes an API for efficiently
```
Some capitalization and grammar tweaks:
```
The real-time I/O or RTIO subsystem exposes an API for efficiently
moving information in a latency and processing friendly
way, between a Zephyr application and external I2C or SPI devices
(such as sensors).
```

dbkinder commented on doc/reference/peripherals/rtio.rst:
```diff
@@ -0,0 +1,49 @@
+********
+
+RTIO provides two significant C structures and APIs for moving bytes around
+efficiently, blocks and pipes.
```
Change this to match the order of the explanations below:
```
efficiently, using pipes and blocks.
```

dbkinder commented on doc/reference/peripherals/rtio.rst:
```diff
@@ -0,0 +1,49 @@
+Potentially leading to other processing steps, perhaps to determine a gravity
+vector, or orientation relative to earth.
+
+RTIO is meant to be useful in not only efficiently moving data around, but
```
Simplify this with:
```
The RTIO subsystem lets you both efficiently move data around and
efficiently process chunks of data at a time.
```

dbkinder commented on doc/reference/peripherals/rtio.rst:
```diff
@@ -0,0 +1,49 @@
+sensors and converting it to SI units with common numerical formats for that
+sensor kind.
+
+Potentially leading to other processing steps, perhaps to determine a gravity
```
Delete the blank line above this so it doesn't start another paragraph and rewrite as:
```suggestion
That could then lead to other processing steps, perhaps to determine a gravity
```

dbkinder reviewed (changes requested):
Some recommended writing edits and we're good to go.

bfrog commented:
@dbkinder all suggestions taken into account

bfrog commented:
This seems to fail sporadically with shippable, I'll look into it

dbkinder reviewed (approved):
doc changes LGTM, thanks!

nvlsianpu commented on include/rtio.h:
```diff
@@ -0,0 +1,479 @@
+/*
+ * Copyright (c) 2019  Thomas Burdick <thomas.burdick@gmail.com>
```
I believe this file should be moved to include/drivers/ directory.

bfrog replied:
I believe your right for master, this is going into a topic branch that likely needs rebasing again to match master

bfrog replied:
I did move this for now. It can be moved back to include/ rather than include/drivers/ if it makes more sense

microbuilder commented:
Hey @bfrog ... just looking through this again, but it sort of makes me think a bit of GPIO in the sense that we actually have two difference GPIO APIs in development:

- [Standard GPIO](https://docs.zephyrproject.org/latest/reference/peripherals/gpio.html)
- Fast GPIO, with the goal is much better performance

While most people realize the limitations of the previous sensor API, I wonder if this isn't moving in a similar direction where we should have `rtio` for a high-throughput API for specialised use cases, but then a more general-purpose API for regular stuff. This does mean moving from one existing API to two new ones, but perhaps the reauirements are different enough that different APIs are justified?

tyrantlizard77 reviewed (dismissed):
where are the syscall handler functions?
In particular, how are you validating the arguments to functions like rtio_block_alloc()?
your test case doesn't run in user mode  either...

alexanderwachter commented on include/rtio.h:
```diff
@@ -0,0 +1,479 @@
+/**
+ * @brief Unused number of bytes in the buffer
+ */
+__syscall u16_t rtio_block_available(struct rtio_block *block);
```
Why do you need a syscall here? do you plan to put the block into "kernel space"?

bfrog replied:
I added __syscalls on suggestion from @andrewboie but I don't think I thought about the implications fully, I've added a more detailed comment about the syscall interface below and question I have about it

alexanderwachter replied:
You should only add __syscall on functions that need the privileges to write or read memory objects that don't belong to the calling user thread. If you have to do so, you must check the provided pointers. See other syscall handlers for example.

bfrog replied:
I've removed this as a syscall, I believe only the rtio_api backed functions (rtio_configure, rtio_trigger) need to be syscalls and thats exactly what I've changed

alexanderwachter commented on include/rtio.h:
```diff
@@ -0,0 +1,479 @@
+/**
+ * @brief Chunk of memory with ownership and byte format versioning
+ */
+struct rtio_block {
```
Did you think about connecting blocks as a list?
Maybe you should have a look at net_buf implementation. Those buffers are really nice to handle.

bfrog replied:
I've more than once considered just using net_buf, I wanted a very specific allocation strategy though and I need a contiguous memory area for dma, I'll look again at using net_buf more directly here as it provides all the niceties of pushing/pulling various primitives already with it I believe

alexanderwachter replied:
It was just a hint. I'm not sure if it makes sense to have the ability to put the buffers into a list.

bfrog replied:
I've changed this API so that blocks may be allocated by any allocation scheme, with a mempool implementation provided by a macro.

bfrog commented:
After discussing with @andrewboie about the syscall interface, I wanted to clarify the intention of this interface and how memory is most likely going to be allocated, freed, and kernel objects created and used.

1. In the user thread a mempool and fifo are statically created with the wrapper types (rtio_pool, rtio_pipe).
2. These are passed into the kernel device driver using rtio_configure, which should be validating them (it does not as of writing this).
3. Kernel device driver allocates and puts blocks allocated from the mempool into the fifo.
4. User thread polls on the fifo, pulls blocks, and frees them back to the pool when done.

Given the above I have some questions regarding syscalls and the general strategy here.

Is creating a k_mempool and a k_fifo in the application and managed by the application unsafe?
Is passing those back to the kernel driver unsafe? What if I can validate the pointers?
Is allocating in the driver and freeing in the application safe?

mbolivar commented on tests/drivers/rtio/rtio_api/src/test_rtio.c:
```diff
@@ -0,0 +1,91 @@
+	int res = rtio_block_alloc(&small_pool, &block, size, K_NO_WAIT);
+
+	zassert_equal(res, 0, "Allocation should not fail");
+	zassert_equal(block->size, size, "Size should be 1024");
```
minus sizeof(struct rtio_block), no?

bfrog replied:
This should be corrected now, as the allocator API accounts for any metadata sizing

mbolivar commented on include/rtio.h:
```diff
@@ -0,0 +1,479 @@
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(u32_t)
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(s8_t)
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(s16_t)
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(s32_t)
```
Do you have any thoughts on 64 bit support?

I'm not asking for 64 bit support right now -- I'm just suggesting it should be seriously considered, since Zephyr is being made 64-bit safe.

bfrog replied:
I added s64/u64 types here

mbolivar commented on include/rtio.h:
```diff
@@ -0,0 +1,479 @@
+{
+	const struct rtio_api *api = dev->driver_api;
+
+	if (!api->configure) {
```
Totally minor, but it's documented that these must be implemented, so you could save a few bytes of code here by eliminating the check.

mbolivar commented on tests/drivers/rtio/rtio_api/src/test_rtio.c:
```diff
@@ -0,0 +1,91 @@
+					   120);
+
+	k_thread_start(source_tid);
+	k_thread_start(sink_tid);
```
Thanks for the test case!

Just to confirm, there's no reason the main thread couldn't be the sink thread, right?

bfrog replied:
Thats correct

mbolivar commented on doc/reference/peripherals/rtio.rst:
```diff
@@ -0,0 +1,48 @@
+9-DoF SI data stream could be used to determine an earth orientation vector
+using sensor fusion.
+
+The RTIO subsystem lets you both efficiently move data around and
```
Thanks for this documentation. It would be nice to add a few sentences about triggers to this.

bfrog replied:
Definitely, I have some function comments on rtio_trigger but nothing in the general docs explaining what its there for.

mbolivar reviewed (commented):
Just a few random comments from a first read-through. I'm going to try to use this on a GPS driver as discussed, then respond in more detail.

mbolivar reviewed (changes requested):
A couple things:

1. I think the wrapper APIs should take account of the size and alignment of a struct rtio_block when sizing and allocating memory pools. The way it currently is (as shown in `test_rtio_pool()`), the user has to account for this themselves. I think this is a sharp edge and makes it harder to write good code with this API.

2. I would really like to see the test cases exercise the meat of the API (configure and trigger); not just alloc/begin/push/end. This will also help document example usages for people like me that are building on top of this.

bfrog commented on include/rtio.h:
```diff
@@ -0,0 +1,479 @@
+	/** Amount of data this buffer can store. */
+	u16_t size;
+
+	/** Start of the buffer data .*/
```
Given the idea now, based on feedback from @andrewboie and thinking harder that it does in fact, make a lot of sense to allow the user to provide their own allocator, I think this needs to be revisited

bfrog commented on include/rtio.h:
```diff
@@ -0,0 +1,479 @@
+ * @brief Chunk of memory with ownership and byte format versioning
+ */
+struct rtio_block {
+	/* Mem block from k_mem_pool */
```
Great, I've added all the appropriate ```**``` as needed I believe

bfrog commented:
A very common request I've gotten on this is "how do I use my sensor data in SI not just the raw bytes" and I think the easiest solution to that problem is to add back the idea of channels and channel descriptors that I had in zio into a rtio_sensor.h header where an rtio_sensor.h is a sensor based on this rtio api, but requires a few additional interfaces be implemented. A rtio_block reader that allows iterating over sensor reading sets (like the set of x,y,z acc/gyro/mag readings from 9DoF sensor) and getting those individual readings back in SI units as fixed point or floating point values.

I will probably propose what that looks like in another commit in this same PR?

mbolivar commented:
> I will probably propose what that looks like in another commit in this same PR?

+1, that would really clear things up for me.

loicpoulain commented on include/drivers/rtio.h:
```diff
@@ -0,0 +1,487 @@
+
+	/** Amount of data this buffer can store. */
+	u16_t size;
+
```
Could we have len and size as u32_t?

jukkar replied:
Using size_t for length/size of the buffer would probably be appropriate here.

bfrog replied:
We most certainly can do size_t's, and I think it makes sense

bfrog replied:
I've switched everything to size_t to start. Self reviewing this I'm not sure I like that the policy describes size in bytes rather than sample sets or values. Especially given my additional work to provide a sensor.h like API for reading data.

pabigot replied:
Offhand I think it makes sense and at this layer is most flexible to use bytes.  Any other measure would complicate/prevent using it for streams of variable-sized observations.  I have some experience with streams of sampled data passing through a transformation system: pretty soon you want to add sample-aligned metadata, and then fork and join nodes that require processing streams of events where each packet carries its own timestamp and internal metadata.

mbolivar replied:
>> I'm not sure I like that the policy describes size in bytes rather than sample sets or values

> Offhand I think it makes sense and at this layer is most flexible to use bytes.

I agree with @pabigot (and your first instinct, @bfrog) on this one. You can always add sensor-specific macros to divide bytes to number of samples.

bfrog replied:
Perfect, will leave it be then

pabigot commented on include/drivers/rtio.h:
```diff
@@ -0,0 +1,486 @@
+ *
+ * The allocator *must* have a memory layout equivalent to
+ * struct rtio_block_allocator
+ */
```
Here and all other public API: document the parameters and the return values, including expected error codes.  Particularly `size` which is presumably in bytes, not blocks (which need not be fixed size).  (A new user of the API will not know and could get confused.)

pabigot commented on include/drivers/rtio.h:
```diff
@@ -0,0 +1,486 @@
+	size_t size;
+
+	/** Start of the buffer data .*/
+	u8_t *_data;
```
If this is always a pointer to the end of the structure, never to an external buffer, maybe it would make sense to have it be something like:

    union {
       u8_t u8[1];
       u32_t alignment; /* or whatever alignment restriction is needed for the data */
    } data;

to avoid the wasted space for a pointer and the code to dereference it.  Would add a slight complication to calculating the actual allocation size.

bfrog replied:
Its not unfortunately. The way I've allowed for different allocators so far is by creating things like 

``` struct rtio_mempool_block {
  struct rtio_block block;
  struct k_mempool_id mempool_id;
} ```

If there's a better way to do this sort of thing in C I'm open to suggestions.

pabigot reviewed (commented):
Some documentation suggestions, some comments with substance.

bfrog commented on include/drivers/rtio.h:
```diff
@@ -0,0 +1,487 @@
+/**
+ * @brief Begin a block read or write
+ */
+static inline void rtio_block_begin(struct rtio_block *block)
```
Excellent, yes, and in fact it would be fantastic if I can get the precision of the clock in the future as in some instances I'd like to interpolate timestamps of sample sets as a future PR will show

bfrog replied:
As the clock source is not yet in, I believe this should be ok until that happens

bfrog commented on include/drivers/rtio.h:
```diff
@@ -0,0 +1,487 @@
+ *
+ * Each Real-Time IO device provides a way to either read, write, or both
+ * blocks of memory allocated from a static pool by the application. The layout
+ * of the each block is understood implicitly by the author. Each device may
```
yes, thats the point of the layout identifier, but its still specific to the source or sink of the block. Each device driver may come up with their own scheme for identifying the memory layout in the block. Its not something that RTIO is meant to specify.

bfrog commented on include/drivers/rtio.h:
```diff
@@ -0,0 +1,487 @@
+ * @file
+ * @brief Real-Time IO device API for moving bytes around fast
+ *
+ * Each Real-Time IO device provides a way to either read, write, or both
```
I like that change, thank you

bfrog commented on doc/reference/peripherals/rtio.rst:
```diff
@@ -0,0 +1,44 @@
+memory as well as an API for sourcing these blocks from a physical device.
+
+Blocks, defined by :c:type:`struct rtio_block`, provide the unit of data
+being transferred. These blocks contain a byte layout identifier provided by
```
Fair enough

bfrog replied:
I've added some descriptions and intended use case for the metadata here

bfrog commented on doc/reference/peripherals/rtio.rst:
```diff
@@ -0,0 +1,44 @@
+representation.
+
+Intermediary processing steps representing both sinks and sources of data are
+also possible. One possible task is taking the raw byte layout coming from a
```
> > I will probably propose what that looks like in another commit in this same PR?
> 
> +1, that would really clear things up for me.

https://github.com/bfrog/zephyr/pull/1 has the beginnings of what I had in mind

bfrog commented:
@mbolivar @microbuilder @pabigot 

An extension of this basic streaming API specific to sensors that I've started on can be found here

https://github.com/bfrog/zephyr/pull/1

I'd greatly appreciate comments early on.

pabigot commented on include/drivers/rtio.h:
```diff
@@ -0,0 +1,487 @@
+		struct rtio_mempool_block *mempool_block =
+			(struct rtio_mempool_block *)(memblock.data);
+		size_t struct_size = sizeof(struct rtio_block);
+		u8_t *dataptr = &((u8_t *)memblock.data)[struct_size];
```
@bfrog Given your comment on how the buffer is not contiguous with the memory immediately after `struct rtio_block` the calculation here looks wrong.  Because it looks like that's what's being stored.  Which overlaps with `mempool_block->id`.  I think.

Maybe the `struct_size` above should be `sizeof(struct rtio_mempool_block)` then.  Which I'd express as:

    u8_t *data_ptr = (u8_t *)memblock.data + sizeof(*mempool_block);

bfrog replied:
yes, thank you, this is an error when switching to allowing any allocator

bfrog replied:
Fixed in the latest

pabigot commented on include/drivers/rtio.h:
```diff
@@ -0,0 +1,486 @@
+
+		mempool_block->id = memblock.id;
+		*block = (struct rtio_block *)mempool_block;
+		rtio_block_init((struct rtio_block *)*block, dataptr, size);
```
The cast here is unnecessary: `*block` is already `struct rtio_block *`.

bfrog commented on include/drivers/rtio.h:
```diff
@@ -0,0 +1,487 @@
+ * byte layout, and timestamps of when a read or write began and ended.
+ */
+struct rtio_block {
+	/** Timestamp in cycles from k_cycle_get_32() marking when a read or
```
So long as there's a common API for obtaining the current time, the precision of the time source, and a simple way to convert to meaningful units of time without storing extra stuff here the commentary can change here to reflect that

bfrog commented:
@pabigot thank you for the thorough review

pizi-nordic commented on include/drivers/rtio.h:
```diff
@@ -0,0 +1,482 @@
+ * @brief Common trigger sources
+ */
+enum rtio_trigger_source {
+GPIO_IRQ,
```
Why this is without indentation?

bfrog replied:
Emacs failed me

pizi-nordic replied:
Use vim ;-)

pizi-nordic commented on tests/drivers/rtio/rtio_api/src/test_rtio.c:
```diff
@@ -0,0 +1,93 @@
+	rtio_block_begin(block);
+	rtio_block_push_u32_t(block, (u32_t)EXPECTED_VAL);
+	rtio_block_end(block);
+	k_fifo_put(fifo, block);
```
From:  https://docs.zephyrproject.org/latest/reference/kernel/data_passing/fifos.html

> FIFO data items must be aligned on a word boundary, as **the kernel reserves the first word of an item for use as a pointer to the next data item in the queue**. 

So the k_fifo_put() will destroy begin_tstamp. Is this intended?

bfrog replied:
100% not intended. Does this mean I need a void* spacer at the beginning of each rtio_block? Is there a better way of doing that?

bfrog replied:
Ok yes, it does, fixed. Not very happy that I need an additional member in rtio block, might reconsider the usage of a fifo over a ringbuffer

pizi-nordic commented on include/drivers/rtio.h:
```diff
@@ -0,0 +1,482 @@
+ * The allocator *must* have a memory layout equivalent to
+ * struct rtio_block_allocator
+ */
+static inline int rtio_block_alloc(void *allocatorv,
```
The `void *` will hide all the errors related to invalid pointer type passed.
Why you are not using `struct rtio_block_allocator *` here?

bfrog replied:
To avoid casting in the case where you've used a specific allocator. I could use a macro here instead if thats more sensible

bfrog replied:
I've fixed this by doing some pointer casting in the macro which defines the mem pool and then always passing in a allocator pointer here now

pizi-nordic commented on include/drivers/rtio.h:
```diff
@@ -0,0 +1,482 @@
+ * associated GPIO device and pin, a counter timeout, or a function call
+ */
+struct rtio_trigger_config {
+	enum rtio_trigger_source source;
```
How the configuration for sensor internal trigger could be added to this struct?

bfrog replied:
In the case of a sensor providing triggers like data ready, fifo watermark met, etc each device
would need to provide its own struct which is then assigned to the void* device_config which is part of rtio_config

pizi-nordic replied:
Then you will have the same issue like here: https://github.com/zephyrproject-rtos/zephyr/pull/17434#discussion_r320730133: The API will be error prone.

bfrog replied:
So possibilities to more immediately resolve this issue are by allowing device drivers to extend the rtio_configuration struct by embedding it as the first member. Its still effectively just a pointer to memory being passed around, but a container_of like macro could be used then, or even just casting would be fine.

In a larger change to this that @mbolivar and I discussed the other day, making the device configurations statically defined in dts and providing an id to the configuration you'd like to use here instead could avoid a few issues. The issue you bring up with the void*, issues @andrewboie brought up about this api being not user-mode safe, and perhaps allowing for some optimization at the driver level.

pabigot replied:
I'm not seeing enough context to understand how or why device configurations would be expressed in devicetree.  Without looking at what this does, but from first principles of a streaming infrastructure, I would expect an RTIO instance to connect somehow to a device that is capable of feeding data.  It's on that device that there would be device-specific API to configure whatever generates an event that indicates that data is ready.

Depending on responsiveness requirements, that event might use `k_poll_signal()` (if RTIO is animated by an external thread) or `k_sem_give()` (if RTIO has a dedicated thread) to indicate data readiness.  Rather like @pizi-nordic has for `k_work_triggered()`[sic] in #14041.  So in short a device that's designed to be an RTIO source should be "pollable" in some sense.

But I'll have to look more closely to be able to make more coherent noises.  I'll add it to the queue.

bfrog replied:
The device stream is pollable by directly polling on the output FIFO.

pizi-nordic commented on include/drivers/rtio.h:
```diff
@@ -0,0 +1,482 @@
+	 *
+	 * Should be set using something like K_MSEC(5) or K_FOREVER
+	 */
+	u32_t time;
```
What is the purpose of this field?

bfrog replied:
In cases where you'd like to meet some time limited capture of data.
Imagine an IMU sensor that does step detection or a touch screen sensor, you'd like to see if there were steps in the last 10 seconds or perhaps touches in the last 10ms, whether or not anything happened. This time is the time limit to capture data before putting the block into the FIFO, empty or not. In the case of K_FOREVER the block would wait to be filled or meet the capture count in the policy.

pizi-nordic replied:
So IMHO `timeout` will be better name. Also, there should be better explanation in the comment.

teburd committed:
rtio: Ramp device and samples for API validation

A ramp rtio driver and samples to validate the API choices in rtio.h,
build common driver functionality, and provide the basis for other
drivers to build on.

Signed-off-by: Tom Burdick <tom.burdick@electromatic.us>
Co-authored-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

teburd committed:
rtio: Context common driver functionality

Adds rtio_context for common driver functionality analagous
to spi_context with SPI. Optionally useful for RTIO implementations
that wish to take advantage of common code for common tasks.

Signed-off-by: Tom Burdick <tom.burdick@electromatic.us>
Co-authored-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

teburd committed:
rtio_sensor: Sensor API using RTIO

Adds a layer of sensor value decoding on top of the RTIO streaming
API to provide physical measurement values in useful numerical
formats and units.

Signed-off-by: Tom Burdick <tom.burdick@electromatic.us>

mbolivar-nordic committed:
fixup! rtio: Real-Time Stream API

mbolivar-nordic committed:
fixup! rtio: Context common driver functionality

tejlmand commented on samples/rtio/ramp_reader/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Before going in, this should be updated to `find_package(Zephyr)`

```suggestion
find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})
```

tejlmand commented on samples/rtio/ramp_throughput/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Please update to find_package(Zephyr)
```suggestion
find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})
```

tejlmand commented on tests/drivers/rtio/rtio_api/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Please update to `find_package(Zephyr)`
```suggestion
find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})
```

tejlmand reviewed (changes requested):
please update to use `find_package(Zephyr)`

bfrog commented:
If I get back to this I'll reopen, but I've personally run out of mental energy on this for the time being, and the product it was being made for is on indefinite hold at the moment

bfrog commented:
@gmarull at this point it needs to be rebased and all the issues fixed, there needs to be automated tests written, and probably most of all a complete working driver put in to showcase the usage. All of those things were in progress but as mentioned in a previous comment for various reasons I've paused on this. I'd be happy to continue working on it, especially if someone is looking to fund the work as it really is not something easily done on the side in my opinion. If someone else wants to take over its here and easily forked, I'd be happy to review and assist in understanding any of my thinking or reasoning.

adamfowleruk commented:
Hi all. We're in the process of finishing a Herald analysis API. One of the future sources of data will be raw sensor data on devices running Zephyr RTOS. This proposed API in this PR would match very well as a 'Source' for our analysis API. Who can I talk to about progressing this Zephyr API please? Happy to help test it out on my range of devices and sensors here.

In particular we will be sourcing temperature, pulse rate/blood ox, ambient light sensors via I2C from an nRF52833. Our mid-level analysis framework (initial use - digital contact tracing exposure risk calculation for COVID-19/Ebola) can be found here for reference: https://github.com/adamfowleruk/herald-for-cpp/blob/0c6d8ba0067f2324cae8563df13834f727577a3a/herald-tests/analysisrunner-tests.cpp#L150

bfrog commented:
@adamfowleruk happy to discuss how we might get this moving forward again, I set it down due to a lack of time as I started other, new projects with other companies.

Happy to pick it up again or help guide others. as I mentioned in previous comments. I know @gmarull was also interested in seeing this move forward.

gmarull commented:
I'd really like to see `rtio` moving forward, unfortunately I can only help on review/testing right now as I'm busy with other things.

avisconti commented:
I'm also interested in see this things getting go again.

henrikbrixandersen commented:
+1 from here as well.

adamfowleruk commented:
I may have time to look at this over the next month. We now have prototype wearable hardware and so will revisit this soon.

bfrog commented:
I'm happy to hear there's interest in this, maybe a first step on my part would be to rebase it to 2.6 and get it compiling

adamfowleruk commented:
@bfrog That'd be great, thanks. If you have any time I'd love to pair with you over zoom to understand what is there now.

teburd commented:
its building again and you can run the ramp_throughput sample, its not PR ready though and might take some time to get there still, happy to discuss on discord

https://github.com/teburd/zephyr/tree/v2.7-rtio
